### PR TITLE
New module: monitoring/grafana_datasource

### PR DESCRIPTION
**Pull request by seuf**
##### SUMMARY
**grafana_datasource** is a new Ansible module that allow you to create/update/delete grafana datasources

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
lib/ansible/modules/monitoring/grafana_datasource.py

##### ANSIBLE VERSION
```
ansible 2.3.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.13 (default, May 10 2017, 20:04:28) [GCC 6.3.1 20161221 (Red Hat 6.3.1-1)]
```

##### ADDITIONAL INFORMATION

This module can manage grafana datasource.
I've only tested it with influxdb and elasticsearch `ds_type`, but it should work with graphite, prometheus and opentsdb too.

Example : 
```
hosts: grafana
tasks:
  - name: creating influxdb datasource
    connection: local
    grafana_datasource:
      name: "influxdb_ansible"
      grafana_url: "http://grafana.company.com"
      grafana_user: "admin"
      grafana_password: "admin"
      org_id: "1"
      ds_type: "influxdb"
      url: "http://influxdb.company.com:8086"
      database: "telegraf"
      state: present
```

### CODE REVIEW

File: lib/ansible/modules/monitoring/grafana_datasource.py [hunk 1]
```diff
@@ -0,0 +1,469 @@
...
+      - opentsdb version (1 for <= 2.1, 2 for ==2.2, 3 for ==2.3)
+    default: 1
+  tsdb_resolution:
+    required: false
+    choices: [1, 2]
+    description:
+      - opentsdb time resolution (1 for second, 2 for millisecond)
+    default: 1
+  validate_certs:
+    required: false
```
Comment (nerzhul) [inline lib/ansible/modules/monitoring/grafana_datasource.py]: missing default value

File: lib/ansible/modules/monitoring/grafana_datasource.py [hunk 2]
```diff
@@ -0,0 +1,469 @@
...
+            tls_ca_cert=dict(type='str', no_log=True),
+            is_default=dict(default=False, type='bool'),
+            org_id=dict(default=1, type='int'),
+            es_version=dict(type='int', default=5, choices=[2, 5]),
+            time_field=dict(default='@timestamp', type='str'),
+            time_interval=dict(type='str'),
+            interval=dict(type='str'),
+            tsdb_version=dict(type='int', default=1, choices=[1, 2, 3]),
+            tsdb_resolution=dict(type='int', default=1, choices=[1, 2]),
+            validate_certs=dict(type='bool')
```
Comment (nerzhul) [inline lib/ansible/modules/monitoring/grafana_datasource.py]: please set default to True

**Review by nerzhul — CHANGES_REQUESTED** [review]
First review, in 1 min

File: lib/ansible/modules/monitoring/grafana_datasource.py [hunk 3]
```diff
@@ -0,0 +1,469 @@
...
+      - Grafana url
+  name:
+    required: true
+    description:
+      - Name of the datasource
+  ds_type:
+    required: true
+    choices: [graphite,prometheus,elasticsearch,influxdb,opentsdb]
+    description:
+     - Type of the datasource C(graphite),C(prometheus),C(elasticsearch),C(influxdb),C(opentsdb)
```
Comment (gundalow) [inline lib/ansible/modules/monitoring/grafana_datasource.py]: No need to repeat the choices.

File: lib/ansible/modules/monitoring/grafana_datasource.py [hunk 4]
```diff
@@ -0,0 +1,469 @@
...
+                  'changed': False,
+                  'id': 0,
+                  'name': data['name']}
+
+    return result
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
```
Comment (gundalow) [inline lib/ansible/modules/monitoring/grafana_datasource.py:443]: Please update as per
https://github.com/ansible/ansible/pull/31930/files

```
required_together=[['grafana_user', 'grafana_password', 'org_id']],
mutually_exclusive=[['grafana_user', 'grafana_api_key']],
```

File: lib/ansible/modules/monitoring/grafana_datasource.py [hunk 5]
```diff
@@ -0,0 +1,469 @@
...
+  grafana_api_key:
+    required: false
+    description:
+      - Grafana API key
+      - If set, grafana_user and grafana_password will be ignored
+  database:
+    required: false
+    description:
+      - Name of the database for the datasource.
+      - This options is required when the ds_type is influxdb or elasticsearch (index name)
```
Comment (gundalow) [inline lib/ansible/modules/monitoring/grafana_datasource.py]: Should be enforced by argspec

File: lib/ansible/modules/monitoring/grafana_datasource.py [hunk 6]
```diff
@@ -0,0 +1,469 @@
...
+    required: false
+    choices: [1, 2, 3]
+    description:
+      - opentsdb version (1 for <= 2.1, 2 for ==2.2, 3 for ==2.3)
+    default: 1
+  tsdb_resolution:
+    required: false
+    choices: [1, 2]
+    description:
+      - opentsdb time resolution (1 for second, 2 for millisecond)
```
Comment (gundalow) [inline lib/ansible/modules/monitoring/grafana_datasource.py]: Can we set the choices to `[seconds, milliseconds]` then convert it before passing the data to grafana?

File: lib/ansible/modules/monitoring/grafana_datasource.py [hunk 7]
```diff
@@ -0,0 +1,469 @@
...
+    required: false
+    default: 5
+    description:
+      - Elasticsearch version (for ds_type = elasticsearch only)
+  time_field:
+    required: false
+    default: timestamp
+    description:
+      - Name of the time field in elasticsearch ds
+      - For example @timestamp
```
Comment (gundalow) [inline lib/ansible/modules/monitoring/grafana_datasource.py]: `For example C(@timestamp).`

File: lib/ansible/modules/monitoring/grafana_datasource.py [hunk 8]
```diff
@@ -0,0 +1,469 @@
...
+  tls_client_key:
+    required: false
+    description:
+      - client tls private key
+      - befins with ----- BEGIN RSA PRIVATE KEY -----
+  tls_ca_cert:
+    required: false
+    description:
+      - tls ca certificate for self signed certificates.
+      - only used when tls_client_cert and tls_client_key are set
```
Comment (gundalow) [inline lib/ansible/modules/monitoring/grafana_datasource.py:105]: required_together?

**Review by gundalow — CHANGES_REQUESTED** [review]
Same comments as in https://github.com/ansible/ansible/pull/31930 apply here

File: lib/ansible/modules/monitoring/grafana_datasource.py (hunk 6 above)
Comment (seuf) [reply lib/ansible/modules/monitoring/grafana_datasource.py]: ok, I've changed the available choices then convert it for grafana api.

File: lib/ansible/modules/monitoring/grafana_datasource.py (hunk 4 above)
Comment (seuf) [reply lib/ansible/modules/monitoring/grafana_datasource.py:443]: I've also added `required_if` for some parameters..

File: lib/ansible/modules/monitoring/grafana_datasource.py (hunk 5 above)
Comment (seuf) [reply lib/ansible/modules/monitoring/grafana_datasource.py]: the `tls_ca_cert` is not mandatory when tls_client auth is enabled.
It's just an other option to trust the certificate of the datasource.

File: lib/ansible/modules/monitoring/grafana_datasource.py (hunk 8 above)
Comment (seuf) [reply lib/ansible/modules/monitoring/grafana_datasource.py:105]: the tls_ca_cert is not mandatory when tls_client auth is enabled.
It's just an other option to trust the certificate of the datasource.

File: lib/ansible/modules/monitoring/grafana_datasource.py (hunk 5 above)
Comment (seuf) [reply lib/ansible/modules/monitoring/grafana_datasource.py]: done with `required_if`

**Comment by gundalow** [comment]
Cut off for new functionality for 2.5 is 15th Jan.

**Comment by seuf** [comment]
Ok. All requested changes are done except for tls_ca_cert option which is not mandatory.

**Comment by seuf** [comment]
Do I need to fix something else to allow this pr to be merged ?

**Comment by seuf** [comment]
OK, I've added the es_version 56 choices and the new max_conccurent_shard_requests option to limit the number of shards requests by elasticsearch.
I've fixed missing documentation, and formating with {} pattern (replaced with `"foo %s bar" % var`).
All checks have passed. 
We are currently using this module in production to automatically create grafana datasources for different organisations. it would be awesome to have it in ansible 2.5.

**Review by gundalow — APPROVED** [review]

**Comment by gundalow** [comment]
Merged into `devel` will be released in `2.5.0`
Thanks!

**Pull request merged**